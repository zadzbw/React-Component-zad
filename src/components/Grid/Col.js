/**
 * Created by zad on 16/12/8.
 */
import React, {Component, PropTypes} from 'react';
import classNames from 'classnames';
import _forEach from 'lodash/forEach';

import './Col.less';

const colPrefix = 'zad-col';

export default class Col extends Component {
  static propTypes = {
    span: PropTypes.number,
    offset: PropTypes.number,
    xs: PropTypes.number,
    sm: PropTypes.number,
    md: PropTypes.number,
    lg: PropTypes.number
  };

  static defaultProps = {
    span: 0,
    offset: 0
  };

  render() {
    const {span, offset, xs, sm, md, lg, className, children, ...props} = this.props;
    const sizeMap = {
      xs: xs,
      sm: sm,
      md: md,
      lg: lg
    };

    let sizeClassObj = {};
    _forEach(sizeMap, (value, key) => {
      sizeClassObj = Object.assign({}, sizeClassObj, {
        [`${colPrefix}-${key}-${value}`]: !!value
      });
    });

    const colClass = classNames(className, {
      [`${colPrefix}-${span}`]: !!span,
      [`${colPrefix}-offset-${offset}`]: !!offset
    }, sizeClassObj);

    return (
      <div {...props} className={colClass}>
        {children}
      </div>
    );
  }
}
