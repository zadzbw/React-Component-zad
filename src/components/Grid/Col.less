@import "../../style/variables";
@import "./GridMixin";

.init-grid-columns();
.make-grid();
.make-grid(-xs);

@media (min-width: @screen-sm-min) {
  .make-grid(-sm);
}

@media (min-width: @screen-md-min) {
  .make-grid(-md);
}

@media (min-width: @screen-lg-min) {
  .make-grid(-lg);
}
