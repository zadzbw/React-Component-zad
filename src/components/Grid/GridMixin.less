@import "../../style/variables";

.init-grid-columns() {
  .col(@index) {
    @item: ~".zad-col-@{index}, .zad-col-xs-@{index}, .zad-col-sm-@{index}, .zad-col-md-@{index}, .zad-col-lg-@{index}";
    .col((@index + 1), @item);
  }

  .col(@index, @list) when (@index =< @grid-columns) {
    @item: ~".zad-col-@{index}, .zad-col-xs-@{index}, .zad-col-sm-@{index}, .zad-col-md-@{index}, .zad-col-lg-@{index}";
    .col((@index + 1), ~"@{list}, @{item}");
  }

  .col(@index, @list) when (@index > @grid-columns) {
    @{list} {
      position: relative;
      min-height: 1px;
      padding-left: (@grid-gap / 2);
      padding-right: (@grid-gap / 2);
    }
  }

  .col(1);
}

.float-grid-columns(@size) {
  .col(@index) {
    @item: ~".zad-col@{size}-@{index}";
    .col((@index + 1), @item);
  }

  .col(@index, @list) when (@index =< @grid-columns) {
    @item: ~".zad-col@{size}-@{index}";
    .col((@index + 1), ~"@{list}, @{item}");
  }

  .col(@index, @list) when (@index > @grid-columns) {
    @{list} {
      float: left;
      flex: 0 0 auto;
    }
  }

  .col(1);
}

.loop-grid-columns(@index, @size) when (@index > 0) {
  .zad-col@{size}-@{index} {
    display: block;
    width: percentage(@index / @grid-columns);
  }

  .zad-col@{size}-offset-@{index} {
    margin-left: percentage(@index / @grid-columns);
  }

  .loop-grid-columns((@index - 1), @size);
}

.loop-grid-columns(@index, @size) when (@index = 0) {
  .zad-col@{size}-@{index} {
    display: none;
  }
}

.make-grid(@size:"") {
  .float-grid-columns(~"@{size}");
  .loop-grid-columns(@grid-columns, ~"@{size}");
}
