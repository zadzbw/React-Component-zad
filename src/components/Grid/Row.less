@import "../../style/variables";
@import "./GridMixin";

.zad-row {
  display: flex;
  flex-flow: row wrap;

  &:before,
  &:after {
    display: flex;
  }

  &-vertical {
    &-top {
      align-items: flex-start;
    }

    &-middle {
      align-items: center;
    }

    &-bottom {
      align-items: flex-end;
    }
  }

  &-horizontal {
    &-start {
      justify-content: flex-start;
    }

    &-end {
      justify-content: flex-end;
    }

    &-center {
      justify-content: center;
    }

    &-space-around {
      justify-content: space-around;
    }

    &-space-between {
      justify-content: space-between;
    }
  }
}
