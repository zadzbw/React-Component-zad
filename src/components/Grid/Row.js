/**
 * Created by zad on 16/12/8.
 */
import React, {Component, PropTypes} from 'react';
import classNames from 'classnames';
import isOneOf from '../../utils/isOneOf';
import Col from './Col';

import './Row.less';

const _type = {
  align: ['top', 'middle', 'bottom'],
  justify: ['start', 'end', 'center', 'space-around', 'space-between']
};
const rowPrefix = 'zad-row';

export default class Row extends Component {
  static propTypes = {
    gap: PropTypes.number,
    align: PropTypes.string,
    justify: PropTypes.string
  };

  static defaultProps = {
    gap: 0,
    align: 'top',
    justify: 'start'
  };

  render() {
    const {gap, align, justify, style, className, children, ...props} = this.props;
    const rowClass = classNames(rowPrefix, className, {
      [`${rowPrefix}-vertical-${align}`]: align && isOneOf(_type.align, align),
      [`${rowPrefix}-horizontal-${justify}`]: justify && isOneOf(_type.justify, justify)
    });
    const rowStyle = gap ? Object.assign({}, {
      marginLeft: gap / -2,
      marginRight: gap / -2,
    }, style) : style;
    const cols = React.Children.map(children, (col) => {
      if (col.type !== Col) {
        // 限制 Row 的子元素只能是 Col
        return null;
      }

      return React.cloneElement(col, {
        style: gap ? Object.assign({}, {
          paddingLeft: gap / 2,
          paddingRight: gap / 2,
        }, col.props.style) : col.props.style
      });
    });

    return (
      <div {...props} className={rowClass} style={rowStyle}>
        {cols}
      </div>
    );
  }
}
