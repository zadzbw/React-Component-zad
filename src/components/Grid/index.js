/**
 * Created by zad on 16/12/8.
 */
import Row from './Row';
import Col from './Col';

export {Row};
export {Col};
