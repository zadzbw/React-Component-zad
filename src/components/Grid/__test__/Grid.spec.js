/**
 * Created by zad on 16/12/9.
 */
import React from 'react';
import {mount} from 'enzyme';
import {mountToJson} from 'enzyme-to-json';
import {Row, Col} from '../../Grid';

describe('Grid test', () => {
  it('normal', () => {
    const wrapper = mount(
      <Row className="test-row" gap={24}>
        <Col className="test-col" span={4}>
          <div className="test-col-box">
            col-4
          </div>
        </Col>
        <Col className="test-col" span={6}>
          <div className="test-col-box">
            col-6
          </div>
        </Col>
        <div className="test-only-col">
          test-only-col
        </div>
      </Row>
    );

    expect(wrapper.find('.test-row').hasClass('zad-row')).toBeTruthy();
    expect(wrapper.find('.test-row').hasClass('zad-row-vertical-top')).toBeTruthy();
    expect(wrapper.find('.test-row').hasClass('zad-row-horizontal-start')).toBeTruthy();
    expect(wrapper.find('.test-row').prop('style').marginLeft).toEqual(-12);
    expect(wrapper.find('.test-row').prop('style').marginRight).toEqual(-12);

    wrapper.find(Col).forEach((col) => {
      expect(col.prop('style').paddingLeft).toEqual(12);
      expect(col.prop('style').paddingRight).toEqual(12);
    });

    expect(wrapper.find('.test-col').at(0).hasClass('zad-col-4')).toBeTruthy();
    expect(wrapper.find('.test-col').at(0).find('.test-col-box').get(0).textContent).toEqual('col-4');
    expect(wrapper.find('.test-col').at(1).hasClass('zad-col-6')).toBeTruthy();
    expect(wrapper.find('.test-col').at(1).find('.test-col-box').get(0).textContent).toEqual('col-6');

    expect(wrapper.find('.test-col')).toHaveLength(2);
    expect(wrapper.find('.test-only-col')).toHaveLength(0);

    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('align and justify', () => {
    const wrapper = mount(
      <Row className="test-row" align={'middle'} justify={'space-around'}>
        <Col className="test-col" span={4}/>
      </Row>
    );

    expect(wrapper.find('.test-row').hasClass('zad-row-vertical-middle')).toBeTruthy();
    expect(wrapper.find('.test-row').hasClass('zad-row-horizontal-space-around')).toBeTruthy();

    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('offset', () => {
    const wrapper = mount(
      <Row className="test-row">
        <Col className="test-col" span={4} offset={8}/>
        <Col className="test-col" span={6} offset={12}/>
      </Row>
    );

    expect(wrapper.find('.test-col').at(0).hasClass('zad-col-offset-8')).toBeTruthy();
    expect(wrapper.find('.test-col').at(1).hasClass('zad-col-offset-12')).toBeTruthy();

    expect(mountToJson(wrapper)).toMatchSnapshot();
  });

  it('responsive', () => {
    const wrapper = mount(
      <Row className="test-row">
        <Col className="test-col" xs={4} sm={6} md={8} lg={12}/>
      </Row>
    );

    expect(wrapper.find('.test-col').hasClass('zad-col-xs-4')).toBeTruthy();
    expect(wrapper.find('.test-col').hasClass('zad-col-sm-6')).toBeTruthy();
    expect(wrapper.find('.test-col').hasClass('zad-col-md-8')).toBeTruthy();
    expect(wrapper.find('.test-col').hasClass('zad-col-lg-12')).toBeTruthy();

    expect(mountToJson(wrapper)).toMatchSnapshot();
  });
});
