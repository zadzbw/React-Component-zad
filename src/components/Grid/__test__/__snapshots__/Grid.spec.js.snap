exports[`Grid test align and justify 1`] = `
<Row
  align="middle"
  className="test-row"
  gap={0}
  justify="space-around">
  <div
    className="zad-row test-row zad-row-vertical-middle zad-row-horizontal-space-around">
    <Col
      className="test-col"
      offset={0}
      span={4}>
      <div
        className="test-col zad-col-4" />
    </Col>
  </div>
</Row>
`;

exports[`Grid test normal 1`] = `
<Row
  align="top"
  className="test-row"
  gap={24}
  justify="start">
  <div
    className="zad-row test-row zad-row-vertical-top zad-row-horizontal-start"
    style={
      Object {
        "marginLeft": -12,
        "marginRight": -12,
      }
    }>
    <Col
      className="test-col"
      offset={0}
      span={4}
      style={
        Object {
          "paddingLeft": 12,
          "paddingRight": 12,
        }
      }>
      <div
        className="test-col zad-col-4"
        style={
          Object {
            "paddingLeft": 12,
            "paddingRight": 12,
          }
        }>
        <div
          className="test-col-box">
          col-4
        </div>
      </div>
    </Col>
    <Col
      className="test-col"
      offset={0}
      span={6}
      style={
        Object {
          "paddingLeft": 12,
          "paddingRight": 12,
        }
      }>
      <div
        className="test-col zad-col-6"
        style={
          Object {
            "paddingLeft": 12,
            "paddingRight": 12,
          }
        }>
        <div
          className="test-col-box">
          col-6
        </div>
      </div>
    </Col>
  </div>
</Row>
`;

exports[`Grid test offset 1`] = `
<Row
  align="top"
  className="test-row"
  gap={0}
  justify="start">
  <div
    className="zad-row test-row zad-row-vertical-top zad-row-horizontal-start">
    <Col
      className="test-col"
      offset={8}
      span={4}>
      <div
        className="test-col zad-col-4 zad-col-offset-8" />
    </Col>
    <Col
      className="test-col"
      offset={12}
      span={6}>
      <div
        className="test-col zad-col-6 zad-col-offset-12" />
    </Col>
  </div>
</Row>
`;

exports[`Grid test responsive 1`] = `
<Row
  align="top"
  className="test-row"
  gap={0}
  justify="start">
  <div
    className="zad-row test-row zad-row-vertical-top zad-row-horizontal-start">
    <Col
      className="test-col"
      lg={12}
      md={8}
      offset={0}
      sm={6}
      span={0}
      xs={4}>
      <div
        className="test-col zad-col-xs-4 zad-col-sm-6 zad-col-md-8 zad-col-lg-12" />
    </Col>
  </div>
</Row>
`;
